### PR TITLE
[VectorCombine] Add foldShuffleToIdentity

### PR DESCRIPTION
**Pull request by davemgreen**
This patch adds a basic version of a combine that attempts to remove shuffles that when combined simplify away to an identity shuffle. For example:
  %ab = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
  %at = shufflevector <8 x half> %a, <8 x half> poison, <4 x i32> <i32 7, i32 6, i32 5, i32 4>
  %abt = fneg <4 x half> %at
  %abb = fneg <4 x half> %ab
  %r = shufflevector <4 x half> %abt, <4 x half> %abb, <8 x i32> <i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
By looking through the shuffles and fneg, it can be simplified to:
  %r = fneg <8 x half> %a

The code tracks each lane starting from the original shuffle, keeping a track of a vector of {src, idx}. As we propagate up through the instructions we will either look through intermediate instructions (binops and unops) or see a collections of lanes that all have the same src and incrementing idx (an identity). We can also see a single value with identical lanes, which we can treat like a splat.

Only the basic version is added here, handling identities, splats, binops and unops. In follow-up patches other instructions can be added such as constants, intrinsics, cmp/sel and zext/sext/trunc.

### CODE REVIEW

**Comment by RKSimon** [comment]
Funnily enough, I have started work on a VectorCombine::foldShuffleOfShuffles fold that will sort of do this (and catch many of these test cases):

 `shuffle(shuffle(x,u,m1),shuffle(y,u,m2)) -> shuffle(x,y,m3)`

But it doesn't recurse, so every fold has to be cost beneficial - are you seeing cases where we need the recursion capability?

**Comment by RKSimon** [comment]
Yes - we've never come up with a good way to accumulate candidate folds and only commit them if/when it becomes cost effective.

Please don't think I want to just take #88743 + #88899 instead of this, I think they're both needed, I just want to make sure we don't end up with a heavy fold constantly doing too much of the combining.

**Comment by davemgreen** [comment]
The trivial cases were just added for testing - many of them are already handled by instcombine. Some of those combines in instcombine are the ones causing the problem, as they already move shuffles through binops without considering the costs and breaking the canonical patterns for interleaving loads.

We were previously trying to fix that in the InterleavedAccess (through replaceBinOpShuffles), but that starts to become impossible as more and more folds happen in the mid-end.

That is separate from this patch though, which shouldn't create the same problems of breaking interleaving patterns without replacing them with identity shuffles, which should always be an improvement.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 1]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+
+  using InstLane = std::pair<Value *, int>;
+
+  auto LookThroughShuffles = [](Value *V, int Lane) -> InstLane {
+    while (auto *SV = dyn_cast<ShuffleVectorInst>(V)) {
+      unsigned NumElts =
+          cast<FixedVectorType>(SV->getOperand(0)->getType())->getNumElements();
+      int M = SV->getMaskValue(Lane);
+      if (M < 0)
+        return {nullptr, -1};
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1568]:
```suggestion
        return {nullptr, PoisonMaskElem};
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 2]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+      } else {
+        V = SV->getOperand(1);
+        Lane = M - NumElts;
+      }
+    }
+    return InstLane{V, Lane};
+  };
+
+  auto GenerateInstLaneVectorFromOperand =
+      [&LookThroughShuffles](const SmallVector<InstLane> &Item, int Op) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1581]:
```suggestion
      [&LookThroughShuffles](ArrayRef<InstLane> Item, int Op) {
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 3]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+        SmallVector<InstLane> NItem;
+        for (InstLane V : Item) {
+          NItem.emplace_back(
+              !V.first
+                  ? InstLane{nullptr, -1}
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1586]:
```suggestion
                  ? InstLane{nullptr, PoisonMaskElem}
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 4]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+      SplatLeafs.insert(Item[0].first);
+      continue;
+    }
+
+    // We need each element to be the same type of value, and check that each
+    // element has a single use.
+    if (!all_of(drop_begin(Item), [&](InstLane IL) {
+          if (!IL.first)
+            return true;
+          if (isa<Instruction>(IL.first) &&
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1634]: Use dyn_cast

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 5]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+    } else if (isa<UnaryOperator>(Item[0].first)) {
+      Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 0));
+    } else {
+      return false;
+    }
+  }
+
+  // If we got this far, we know the shuffles are superfluous and can be
+  // removed. Scan through again and generate the new tree of instructions.
+  std::function<Value *(const SmallVector<InstLane> &)> generate =
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1657]:
```suggestion
  auto Generate =
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 6]
```diff
@@ -1547,6 +1548,148 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+      [&](const SmallVector<InstLane> &Item) -> Value * {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1658]:
```suggestion
      [&](ArrayRef<InstLane> Item) -> Value * {
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 7]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+                  ? InstLane{nullptr, PoisonMaskElem}
+                  : LookThroughShuffles(
+                        cast<Instruction>(V.first)->getOperand(Op), V.second));
+        }
+        return NItem;
+      };
+
+  SmallVector<InstLane> Start;
+  for (unsigned M = 0; M < Ty->getNumElements(); ++M)
+    Start.push_back(LookThroughShuffles(&I, M));
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1595]:
```suggestion
  SmallVector<InstLane> Start(Ty->getNumElements());
  for (unsigned M = 0, E = Ty->getNumElements(); M < E; ++M)
    Start[M]= LookThroughShuffles(&I, M);
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 8]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+    // element has a single use.
+    if (!all_of(drop_begin(Item), [&](InstLane IL) {
+          if (!IL.first)
+            return true;
+          if (auto *I = dyn_cast<Instruction>(IL.first); I && !I->hasOneUse())
+            return false;
+          return IL.first->getValueID() == Item[0].first->getValueID() &&
+                 (!isa<IntrinsicInst>(IL.first) ||
+                  cast<IntrinsicInst>(IL.first)->getIntrinsicID() ==
+                      cast<IntrinsicInst>(Item[0].first)->getIntrinsicID());
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1639]:
```suggestion
          auto *I = dyn_cast<Instruction>(IL.first);
          if (I && !I->hasOneUse())
            return false;
          return IL.first->getValueID() == Item[0].first->getValueID() &&
                 (!I || I->getIntrinsicID() ==
                      cast<IntrinsicInst>(Item[0].first)->getIntrinsicID());
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 9]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+      Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 0));
+    } else {
+      return false;
+    }
+  }
+
+  // If we got this far, we know the shuffles are superfluous and can be
+  // removed. Scan through again and generate the new tree of instructions.
+  std::function<Value *(ArrayRef<InstLane>)> generate =
+      [&](ArrayRef<InstLane> Item) -> Value * {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1657]:
```suggestion
  auto Generate =
      [&](ArrayRef<InstLane> Item) -> Value * {
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 10]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+    if (IdentityLeafs.contains(Item[0].first) &&
+        all_of(drop_begin(enumerate(Item)), [&](const auto &E) {
+          return !E.value().first || (E.value().first == Item[0].first &&
+                                      E.value().second == (int)E.index());
+        })) {
+      return Item[0].first;
+    } else if (SplatLeafs.contains(Item[0].first)) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1664]:
```suggestion
    if (SplatLeafs.contains(Item[0].first)) {
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 11]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+        Builder.SetInsertPointPastAllocas(I.getParent()->getParent());
+      SmallVector<int, 16> Mask(Ty->getNumElements(), Item[0].second);
+      return Builder.CreateShuffleVector(Item[0].first, Mask);
+    }
+
+    auto *I = cast<Instruction>(Item[0].first);
+    SmallVector<Value *> Ops;
+    unsigned E = I->getNumOperands();
+    for (unsigned Idx = 0; Idx < E; Idx++)
+      Ops.push_back(generate(GenerateInstLaneVectorFromOperand(Item, Idx)));
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1677]:
```suggestion
    SmallVector<Value *> Ops(I->getNumOperands());
    for (unsigned Idx = 0, E = I->getNumOperands(); Idx < E; Idx++)
      Ops[Idx] = Generate(GenerateInstLaneVectorFromOperand(Item, Idx));
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 12]
```diff
@@ -1547,6 +1548,147 @@ bool VectorCombine::foldShuffleOfCastops(Instruction &I) {
...
+    Builder.SetInsertPoint(I);
+    if (auto BI = dyn_cast<BinaryOperator>(I))
+      return Builder.CreateBinOp((Instruction::BinaryOps)BI->getOpcode(),
+                                 Ops[0], Ops[1]);
+    if (auto UI = dyn_cast<UnaryOperator>(I))
+      return Builder.CreateUnOp((Instruction::UnaryOps)UI->getOpcode(), Ops[0]);
+    llvm_unreachable("Unhandled instruction in generate");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1684]:
```suggestion
    assert(isa<UnaryOperator>(I) && "Expected either Unary or Binary Op.");
    return Builder.CreateUnOp((Instruction::UnaryOps)I->getOpcode(), Ops[0]);
```

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 13]
```diff
@@ -192,11 +164,9 @@ define <8 x half> @splatandidentity(<8 x half> %a, <8 x half> %b) {
...
 define <8 x half> @splattwice(<8 x half> %a, <8 x half> %b) {
 ; CHECK-LABEL: @splattwice(
-; CHECK-NEXT:    [[AS:%.*]] = shufflevector <8 x half> [[A:%.*]], <8 x half> poison, <4 x i32> zeroinitializer
-; CHECK-NEXT:    [[BS:%.*]] = shufflevector <8 x half> [[B:%.*]], <8 x half> poison, <4 x i32> zeroinitializer
-; CHECK-NEXT:    [[AB1:%.*]] = fadd <4 x half> [[AS]], [[BS]]
-; CHECK-NEXT:    [[AB2:%.*]] = fadd <4 x half> [[AS]], [[BS]]
-; CHECK-NEXT:    [[R:%.*]] = shufflevector <4 x half> [[AB1]], <4 x half> [[AB2]], <8 x i32> <i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
+; CHECK-NEXT:    [[TMP1:%.*]] = shufflevector <8 x half> [[B:%.*]], <8 x half> poison, <8 x i32> zeroinitializer
+; CHECK-NEXT:    [[TMP2:%.*]] = shufflevector <8 x half> [[A:%.*]], <8 x half> poison, <8 x i32> zeroinitializer
+; CHECK-NEXT:    [[R:%.*]] = fadd <8 x half> [[TMP2]], [[TMP1]]
```
Comment (RKSimon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:169]: This looks like we might need a foldBinopOfShuffles as well......

**Review by RKSimon — COMMENTED** [review]
Do you intend to support intrinsics at some point? I notice you have a abs test

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 14]
```diff
@@ -1667,6 +1668,149 @@ bool VectorCombine::foldShuffleOfShuffles(Instruction &I) {
...
+                                      E.value().second == (int)E.index());
+        })) {
+      IdentityLeafs.insert(Item[0].first);
+      continue;
+    }
+    // Look for a splat value.
+    if (all_of(drop_begin(Item), [&](InstLane &IL) {
+          return !IL.first ||
+                 (IL.first == Item[0].first && IL.second == Item[0].second);
+        })) {
```
[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1744]: We don't currently handle undef zero'th lanes due to the complexities it causes throughout the combine. It is the `if (!Item[0].first)` above line that checks if the first element is undef.

**Review by davemgreen — COMMENTED** [review]
> Do you intend to support intrinsics at some point? I notice you have a abs test

Yep, those and a few other operations. I wanted to make sure the basics were working before adding the others but they are hopefully relatively straight forward.

**Comment by davemgreen** [comment]
Rebased and ping. Thanks

**Review by RKSimon — APPROVED** [review]
Final question - do we need handling for DivRem instructions to make sure we don't introduce poison mask elements? see 4cc9c6d98dfef90d1ffa69977d13ffa2894a10f7

**Comment by davemgreen** [comment]
> Final question - do we need handling for DivRem instructions to make sure we don't introduce poison mask elements? see [4cc9c6d](https://github.com/llvm/llvm-project/commit/4cc9c6d98dfef90d1ffa69977d13ffa2894a10f7)

Thanks - I hadn't considered div/rem as they don't as often come up for vectors. I can exclude them for now if that sounds OK, but as far as I understand they might be fine as we end up removing the shuffles so do not have poison lanes left over.
https://alive2.llvm.org/ce/z/NAyQty

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by dyung** [comment]
Hi @davemgreen, we are seeing an assertion failure in a few internal tests which I bisected back to your change. I have filed #91078 for the issue, can you take a look?

**Comment by davemgreen** [comment]
Thanks for the heads up - I'll take a look.

**Comment by davemgreen** [comment]
Hi - yeah I'll take a look. Thanks for letting me know.

**Pull request merged**